### ISSUE TITLE
With OpenSSL 3, net-ssh 7 cannot use private keys provided through start options of ssh config file

### ISSUE THREAD

**Issue by guss77** [net-ssh/net-ssh#874]
### Expected behavior

With net-ssh 7.0.1 and openssl 3.0.2, the following code should be able to connect to the remote SSH server:

```
Net::SSH.start(host, user, { keys: [ 'key.pem' ] })
```

### Actual behavior

`Net::SSH.start` fails to connect and throws `Net::SSH::AuthenticationFailed` exception with the message `Authentication failed for user username@hostname`. When setting the option `verbose: Logger::DEBUG`, I get these messages:

```
E, [2022-06-30T12:00:10.511537 #1037434] ERROR -- net.ssh.authentication.key_manager[76c]: could not load private key file `key.pem': OpenSSL::PKey::PKeyError (rsa#set_key= is incompatible with OpenSSL 3.0)
E, [2022-06-30T12:00:10.511572 #1037434] ERROR -- net.ssh.authentication.session[730]: all authorization methods failed (tried none, publickey)
```

### System configuration

- net-ssh version 7.0.1
- Ruby version 3.0.2p107

### Example App

```ruby
#!/usr/bin/env ruby

gem 'net-ssh', '=7.0.1'

require 'net/ssh'

host = 'geek.co.il'
user = 'ubuntu'
opts = { keys: [ '~/.ssh/private/aws-vhost.pem' ], verbose: Logger::DEBUG }
Net::SSH.start(host, user, opts) do |ssh|
	puts ssh.exec!('cat /etc/machine-id')
end
```

**Comment by guss77** [comment]
Here's a video showing the behavior of the test app vs. openssh client

https://user-images.githubusercontent.com/381782/176640617-56763566-e633-47ef-894c-545f11aa99d8.mp4

**Comment by schopin-pro** [comment]
Following up on @dentarg 's comment on the other issue, to implement the RSA private key support in the same way as it's been done for the public key, you'll need to create an ASN.1 representation of the key. The best resource I could find that describes that representation (or rather, those representations, as it depends whether we're using PKCS#1 or PKCS#8) is this blog article:

https://www.thedigitalcatonline.com/blog/2018/04/25/rsa-keys/

**Comment by nikita015** [comment]
I'm also getting 
```
.rvm/gems/ruby-3.1.2/gems/net-ssh-7.0.1/lib/net/ssh/authentication/key_manager.rb:168:in `rescue in sign': the given identity is known, but the private key could not be loaded: OpenSSL::PKey::PKeyError (rsa#set_key= is incompatible with OpenSSL 3.0) (Net::SSH::Authentication::KeyManagerError)                                                               
.rvm/gems/ruby-3.1.2/gems/net-ssh-7.0.1/lib/net/ssh/buffer.rb:266:in `set_key': rsa#set_key= is incompatible with OpenSSL 3.0 (OpenSSL::PKey::PKeyError)
```

Wasn't this fixed in 7.0.0-beta?

**Comment by guss77** [comment]
I believe this was the relevant change that introduced the issue: https://github.com/ruby/openssl/commit/6848d2d969d90e6a400d89848ecec21076b87888

As far as I can tell, the `EVP_PKEY_fromdata()` method promised by rhenium had yet to materialize.

**Comment by schopin-pro** [comment]
I think you misunderstood my remark :).

I'm one of the authors of the patchset porting the code to OpenSSL 3.0,
and to work around the limitations of the new API we've basically ended
up encoding all key data in an in-memory ASN.1 representation before
feeding it wholesale to libssl. It was a PITA to find for each key the
exact format that would be understood by libssl, hence my comment
pointing to a resource for whoever is willing to tackle this issue.

**Comment by nikita015** [comment]
Now I got it, thanks for clarifying @schopin-pro

**Comment by helperton** [comment]
For those needing a workaround until this issue is addressed, this helped me: https://stackoverflow.com/questions/72818958/ruby-net-ssh-7-0-1-ubuntu-22-openssl-3-issue

**Comment by alik78** [comment]
Hi! I am still getting this Authentication::KeyManagerError
```
Net::SSH::Authentication::KeyManagerError: the given identity is known, but the private key could not be loaded: OpenSSL::PKey::PKeyError (rsa#set_key= is incompatible with OpenSSL 3.0)
```

I have this env:
``` 
❯ ruby -v
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [x86_64-darwin21]
❯ gem list | grep net-ssh
net-ssh (7.1.0.beta1, 7.0.0.beta1)
```

Do you know when this issue might be resolved? Thanks!

**Comment by corasaurus-hex** [comment]
I managed to fix this problem by first converting the openssh private key to an RSA private key:

**WARNING**: This will change your private key file in-place, make a backup of it. Just hit enter at both prompts if you don't want a password set.

```
ssh-keygen -p -f /path/to/my/private.key -m PEM
```

And then I add `pubkey_algorithms` as a valid option:

```ruby
require "net/ssh"

module Net::SSH
  VALID_OPTIONS << :pubkey_algorithms
end
```

And then connect using:

```ruby
Net::SFTP.start(endpoint, user, {key_data: [my_private_key_read_from_disk], pubkey_algorithms: ['ssh-rsa'], .....})
```

YMMV but I just got this working.

**Comment by dvodvo** [comment]
@corasaurus-hex I see why this solution can be a thumbs up.  

Here is where I scratch my head: that user is already successfully deploying other applications to the same server with net-ssh (7.0.1) and sshkit (1.21.3) (per Gemfile.lock) but with ruby 2.6.1, executes ssh and scp to that server, and must continue doing so? also [documented this matter here](https://stackoverflow.com/questions/75703800/rails-deploy-command-fails-rsaset-key-being-incompatible-with-openssl-3-0).

**Comment by mfazekas** [comment]
This should be fixed in 7.1.0
